### ISSUE TITLE
.NET entry point and deployment API

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#3499]
Here's some API feedback from me. Please take it as just one person's take, not mandating we address them. I'll also add issues as I encounter them (is it worth doing an actual API review?)

* Why `Deployment.RunAsync`, when there is no alternative `Run` method? Should we simply call this `Run`, and assume all deployment are asynchronous? Is this a .NET API convention?
* Why does `RunAsync` return an `int`? The docs don't explain and I ended up three levels deep in the async runner code when I tried to figure it out ... What do we want this return to mean?
* Related to all of this, I would have expected the opening stanza to simply be
    ```csharp
    async void Main() {
        await Deployment.Run(() => ...);
    }
    ```
* Some of the generated type names are exceedingly verbose. (My favorite is `Aws.S3.Inputs.BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDe    faultArgs`). Is it possible to make these work with anonymous types? When I tried, I got some error that an anonymous type couldn't be implicitly converted to an `Output<...>`.

**Comment by mikhailshilkov** [comment]
I can't find a fresh official doc saying so, but [it seems](https://github.com/dotnet/corefx/issues/31275#issuecomment-407181532) that the `Async` suffix is still the convention to follow.

Regarding the `int`, maybe we should set the result with something like `Environment.Exit(code)` and migrate to `async void Main`.

Ideally, I would look into removing the explicit call to `Deployment.RunAsync`.

**Comment by CyrusNajmabadi** [comment]
> Should we simply call this Run, and assume all deployment are asynchronous? Is this a .NET API convention?

Yes.  The method is task returning, and thus should end with `Async`.

> Why does RunAsync return an int? The docs don't explain and I ended up three levels deep in the async runner code when I tried to figure it out ... What do we want this return to mean?

Because if we don't fail your app, then pulumi will think your program succeeded and your program may complete successfully even though you failed.

> Related to all of this, I would have expected the opening stanza to simply be 

We need a way to represent failure and to signal that to pulumi.  Note: we could potentially do that by hard exiting with Environment.Exit().  I'll have to see if htat could work.

> Is it possible to make these work with anonymous types? 

No, it is not.  

We can def try to reduce these names, and it is on our list of things to improve.

**Comment by CyrusNajmabadi** [comment]
> async void Main

Note: this will not work.  It doesn't even compile.  

This is because C# needs the actual *Task* of the Main method so it can generate:

```c#
private static void Main(string[] args)
    => UserMain(args).GetAwaiter().GetResult();

private static async Task UserMain(string[] args)
...
```

**Comment by CyrusNajmabadi** [comment]
Note: one thing we can definitely do is to move away from expecting the c# app to actually be... well... an app.  Instead it can just be a dll where you write something like this:

```c#
using Pulumi;

[Stack]
class MyStack : Stack
{
    public override Task RunAsync()
    {
        // .. your code ..
    }
}
```

However, in order to support this we would actually need a stub .net program that we called into that effectivley contained the current `Main + Deployment.Run` code.  It would load the library, use some DI framework (likely `Mef`) and then locate and call into this attributed code.

That would be the most .net-y way of doing things.  However, it will require building new pieces of infrastructure and will involve a more complex loading/execution model.  So it didn't make the cut for hte preview.

**Comment by joeduffy** [comment]
> Because if we don't fail your app, then pulumi will think your program succeeded and your program may complete successfully even though you failed.

If you throw an exception, doesn't that result in failure? I guess I don't even know where this `int` is coming from ... The user code doesn't produce it, right?

**Comment by joeduffy** [comment]
> No, it is not.

Can you explain in a bit more detail? I 100% believe you, obviously, but would love to be educated more on this. I assume it's some interaction with anonymous types and implicit conversions?

**Comment by joeduffy** [comment]
> Note: this will not work. It doesn't even compile.
>
> This is because C# needs the actual Task of the Main method so it can generate:

Fascinating. Since C# supports void async methods, I just assumed the compiler would do the transformation for you. I see that it does not.

Amended, I would personally expect:

```
static async Task Main() {
    await Deployment.Run(() => ...);
}
```

or, as Mikhail alluded to, maybe even just:

static async Task Main() {
    ...;
}

although I appreciate that this may not be possible without a custom host or something.

**Comment by lukehoban** [comment]
> without a custom host 

This is the direction I believe we are planning on going - as @CyrusNajmabadi noted in https://github.com/pulumi/pulumi/issues/3499#issuecomment-553097414.

**Comment by CyrusNajmabadi** [comment]
> Can you explain in a bit more detail? I 100% believe you, obviously, but would love to be educated more on this. I assume it's some interaction with anonymous types and implicit conversions?

Sure :)

Sorry for the brevity before.  Being on a phone isn't pleasant for this :)

Basically, anonymous types have no real effective mechanism to tie them to real types in a static fashion.  THe compiler treats them effectively as their own distinct types without any way to 'bind/connect' them to any sort of real type.  You can try to have some sort of system bridge between the two for you (effectively by using reflection/dynamic), but you effectively then get *zero* help/understanding from the compiler/IDE about what you're doing.  i.e. you'd have to type everything as object/dynamic just to use the anonymous types, and you'd basically be SOL in terms of figuring out what actual property names to use (and what types they were) when writing your program.

Anonymous types were the way to try to generate small little types only within the scope of a method.  Or for use in heavily dynamic scenarios where not having any compiler support would be acceptable (i.e. dynamically mapping an object in C# to columns in a db table).  They very much sit apart from the rest of hte static types people would want to use.

**Comment by CyrusNajmabadi** [comment]
> Fascinating. Since C# supports void async methods, I just assumed the compiler would do the transformation for you. I see that it does not.

So, 'async void' methods are really only intended for event-firing, and only to tie the previous way .net did things around events to the new async world.  Very specifically, they suck for almost ll parties as they literally mean (since hte method is void) you can get no insight on what the async work actually is.

Presumably the compiler could have actually seen `async void` and actually just gone "actually... no... we'll compile that to `async Task`", but that was seen as likely to just be confusing to users as it wouldn't be like any other `async void` method anywhere else.  `Main` is potentially special enough to warrant it, but that's not what the LDM did :)

**Comment by CyrusNajmabadi** [comment]
> If you throw an exception, doesn't that result in failure? I guess I don't even know where this int is coming from ... The user code doesn't produce it, right?

So here's the general idea:

Pulumi recognizes three ways a user program can exit:

1. Success.  The program exits with 0 and we proceed with actually performing all the changes.
2. Failure with a message reported to the user through the right channels (i.e. through the diagnostic subsystem).  This happens through a user-crash, or many of our *special* exceptions that send messages, but do not print stack traces.  In this case, because we reported things properly to the user, we also want to suppress messages *additional* error-handling messages higher up the stack.
3. Failure in the host itself outside of the user's code.  

'2' is especially important.  Consider the case of the user missing a config variable.  We want to stop the user app immediately, but we don't want things like a callstack, or additional unnecessary scary messages talking about failure.  Because we know we reported the missing config variable.  We can give a great experience.

The way we distinguish '2' and '3' (in nodejs and dotnet) is through a special error code we return.

**Comment by CyrusNajmabadi** [comment]
> If you throw an exception, doesn't that result in failure?

Right.  But how do we distinguish:

1. user exceptions in their code?
2. catastrophic exceptions in *our* code?
3. pleasant exceptions that indicate a change that can happen, that shouldn't come across as a crash to the user?

etc. etc.

Note: if we move to a host+mef-dll model, this *could* go away.  Effectively, the custom host exe would literally look like this:

```c#
public async Task Main(string[] args)
{
    // Pseudo code
    var userStack = MefContainer.Find<Stack>();
    if (userStack == null) throw ...

    return Deployment.Instance.RunAsync(async () =>
    {
        await userStack.RunAsync().ConfigureAwait(false);
        return GetExports(userStack);
    });
}
```

Basically, teh user will not have to worry about things like the `int` or `Main` or calling `Deployment.Instance.RunAsync` themselves.  We'll do all of that on their behalf, and they can just have a simple class with some nice attributes to give us the info we need to call into their code.

--

What we have now is a starting point to ensure we have something that works.  We can definitely improve on this safely in the near term.

Note: i don't believe you need me for this either :)  Anyone moderately comfortable with .NET could add support for the custom host.  It would take a little bit of learning about MEF.  But that should be not a big deal :)

**Comment by joeduffy** [comment]
> This is the direction I believe we are planning on going - as @CyrusNajmabadi noted in #3499 (comment).

What I was suggesting was slightly different, in that you'd still write an ordinary Main entrypoint. More like [STAThread]. Maybe just semantics, but the idea being you just write a Main and don't need to do anything funny like subclass a special type which introduces new concepts. Not sure if that's possible without a custom host (which afaik needs to be written in C++).

**Comment by joeduffy** [comment]
> Main is potentially special enough to warrant it, but that's not what the LDM did :)

Ah got it, that would have made intuitive sense to me, but if that's not how things work, then that's that!

**Comment by mikhailshilkov** [comment]
While creating a host, we should take other scenarios into consideration, like running PowerShell/C#/F# scripts.

Are there downsides to the pluggable DLL approach? Does this make dependency management more complicated/increase the chances of version conflicts? E.g. Azure Functions, who use the "host + DLL with attributes" model, brought a huge pile of binding-redirect-related issues and can't solve it up until now.

**Comment by mikhailshilkov** [comment]
We could keep using a console application but invent any kind of simple `Main` method. For instance, introduce the attribute-based stacks as @CyrusNajmabadi had above, implement something like

```cs
public static int Main() => Pulumi.Run();
```

and then `Pulumi.Run` would discover stacks and do the magic. 

An overload could accept a script file, or a PowerShell file, or what not. If needed, we could create our own builder in line with what ASP.NET does:

```cs
public static void Main()
{
    var host = new WebHostBuilder()
        .UseKestrel()
        .UseContentRoot(Directory.GetCurrentDirectory())
        .UseIISIntegration()
        .UseStartup<Startup>()
        .Build();

    host.Run();
}
```

The user would still be in charge of running the `Main` and controlling the complete dependency list to enable any advanced scenario.

**Comment by CyrusNajmabadi** [comment]
So, we basically have to decide how far we want to go here.  And there are several avenues open to us.  For exapmle, there are strong arguments to be made that not only should we have:

`pulumi update ...` but we should also have:

`dotnet pulumi update ...`

i.e. `dotnet` is a pluggable system.  

For now, my preference though would be to stick with our consistent model for everything.  user just invokes `pulumi` and `pulumi` can appropriately do what's necessary to launch the .NET support and execute hte user code.

I do think it's *not* .NET'y to have an executable launched by another executable which has to trampoline into things.  I think it would make more sense for people that they just deliver a library with the right attributes.  This feels very much like a normal .NET plugin using attributes/MEF as the common discovery model.  

Our current system blends the host/user-code into one thing, and it demonstrates that .net can work.  but i think we really want the custom host to be something we own and ship, and we leave it to users to just be a class+mef-attributes to work like most of hte ecosystem out there to start with.

--

Later on, we can see if we want to invert things even more and become just a `dotnet` plugin.  But i think that could be in a post 1.0 world for us.

**Comment by CyrusNajmabadi** [comment]
> I am still not following entirely... we may do this internally but the developer isn't exposed to this detail. 

The developer *currently* supplies hte Main entrypoint to their user program.  And we need the return value of that to tell the Pulumi language host about this special error code.  i.e. the path in/out is:

```
Pulumi Language Host ->
Launches User Program ->
Which enters `Task<int> Main` ->
Where the user calls `Deployment.Instance.Run` ->
Passing in their lambda.

If `Deployment.Instance.Run` encounters a problem it can then pass the proper
return value back out.
```

> The way we do it in the user-facing part of Node.js runtime is a special exception type, isn't it?

Yes.  But both in Node.js and .NET we have to catch that exception *in* the program and pass that to the language host. 

The difference *right now* is that our SDK has a special *Node.js* program we've written that itself then loads the user program and does the try/catch and returns the right value.

We don't have that *today* for .Net.  Adding such a thing has been covered in previous posts (including https://github.com/pulumi/pulumi/issues/3499#issuecomment-553577379).  

In such a world, it would be like this:

```
Pulumi Language Host ->
Launches Pulumi-Host Program -> //<-- this is new
Which enters `Task<int> Main` -> 
Which uses MEF to find and load the user plugin -> 
Which calls `Deployment.Instance.Run` ->
Passing in the entrypoint to the user's actual code.
```

In this system we own the entire stack *up until the user's actual code*.  Then we can hide all these details.  As mentioned though, we do not have that *today*. It was simply out of scope for the .NET alpha we were building.  It would have required creating this host, determining a way to load the user code (likely MEF), ensuring it loaded properly and whatnot.  

I would like us to build this for the real 1.0 sdk though.  The alpha we have demonstrates that the core tech works.  But 1.0 can focus on these sorts of things. 

Note: this is covered already in our future work issue: https://github.com/pulumi/pulumi/issues/3470

**Comment by mikhailshilkov** [comment]
I'm tearing #3470 apart, so I'm going to use this issue to track our entry point / hosting model discussion

**Comment by damianh** [comment]
Great discussion - I'll like to add some feedback to this issue instead of creating another (if irrelevant, let me know).

As it stands, this is an app with a `Main()` and the dotnet way to run these is `dotnet run` or `app.exe` if published as a self contained deployment. That's what I expect anyway. If it is to be a plugin dll to another host, then the it I think it should be a class library.

What I found jarring from this experience is that if I open a console application project in VisualStudio/Rider, I fully expect to be able to F5 and debug/step through it successfully. I would not expect to have to go to CLI to run a different program to run my program. I am really missing the ability to debug a `pulumi preview`.

I understand the entire Pulumi model is premised around "User runs Pulumi" -> "Pulumi runs user code" but what I _think_ I'd prefer is "User runs their code" -> "User code runs Pulumi".

**Comment by mikhailshilkov** [comment]
@damianh Thanks for your feedback, appreciated!

There are several aspects here. In reverse order:

1. We see many users that expect to "just run" a Pulumi program and see it work. Unfortunately, this is not compatible with the way Pulumi is engineered: the workflow relies on the Pulumi CLI being the driver and changing it to the other way around is non-trivial. We are aware about this mismatch and need to do a better job about educating newcomers. We'll also take this into consideration for long term, but the flow is not likely to change any time soon.

2. This doesn't mean that debugging is not possible. It is possible to attach an IDE to a running Pulumi program instantiated from the CLI (maybe add a `while (true)` in the beginning to make timing easy). We track the first-class debugger support in #1372, which we hope to solve for 2.0.

3. Currently, the decision is to stick with `exe` build artifacts and make further improvements to how `Main` is defined and how it integrates with the rest of the .NET Core ecosystem.

**Comment by damianh** [comment]
Thanks for the response @mikhailshilkov 
1. Yes I understands that is how things stand today; long term consideration 👍 
2. Look forward to first class support.
3. Ok (I susspect this will be an ongoing topic anyway).

Cheers

**Comment by mikhailshilkov** [comment]
I think we agreed to use the `Stack` component as the blessed way to define .NET programs. The component is implemented and released in 1.9.0, so I'll start transitioning templates and examples shortly.
